File: networkmanager.cpp
#include "dbus.h"
#include "networkmanager.h"

NetworkManager::NetworkManager(QObject* parent)
  : QObject(parent)
{
  setObjectName("NetworkManager");
}

QVariantMap getAccessPointInfo(const QDBusObjectPath& path) {
  QVariantMap m = getProperties(path, "org.freedesktop.NetworkManager.AccessPoint");
  if (m.contains("Ssid"))
    m.insert("Ssid", QString(m["Ssid"].toByteArray()));
  return m;
}

QVariant getAccessPoints(const QDBusObjectPath& path) {
  QDBusInterface i("org.freedesktop.NetworkManager", path.path(),
                   "org.freedesktop.NetworkManager.Device.Wireless",
                   QDBusConnection::systemBus());
  QDBusReply<QList<QDBusObjectPath>> r = i.call("GetAccessPoints");
  if (r.isValid()) {
    QVariantList l;
    QListIterator<QDBusObjectPath> i(r.value());
    while (i.hasNext())
      // l << i.next().path();
      l << getAccessPointInfo(i.next());
    return l;
  } else {
    return dbusError(r.error());
  }
}

QVariant getDeviceInfo(const QDBusObjectPath& path) {
  QVariantMap m(getProperties(path, "org.freedesktop.NetworkManager.Device"));
  m = mergeProperties(m, path, "org.freedesktop.NetworkManager.Device.Wired");
  m = mergeProperties(m, path, "org.freedesktop.NetworkManager.Device.Wireless");
  m = mergeProperties(m, path, "org.freedesktop.NetworkManager.Device.Cdma");
  m = mergeProperties(m, path, "org.freedesktop.NetworkManager.Device.Gsm");
  m = mergeProperties(m, path, "org.freedesktop.NetworkManager.Device.Serial");
  m = mergeProperties(m, path, "org.freedesktop.NetworkManager.Device.Bluetooth");
  m = mergeProperties(m, path, "org.freedesktop.NetworkManager.Device.OlpcMesh");
  m = mergeProperties(m, path, "org.freedesktop.NetworkManager.Device.Modem");

  QVariant acc(getAccessPoints(path));
  if (!acc.toMap().contains("error"))
    m.insert("AccessPoints", acc);
  return m;
}

QVariant NetworkManager::getDevices() {
  QDBusInterface i("org.freedesktop.NetworkManager",
                   "/org/freedesktop/NetworkManager",
                   "org.freedesktop.NetworkManager",
                   QDBusConnection::systemBus());
  QDBusReply<QList<QDBusObjectPath>> r = i.call("GetDevices");
  if (r.isValid()) {
    QVariantList l;
    QListIterator<QDBusObjectPath> i(r.value());
    while (i.hasNext())
      l << getDeviceInfo(i.next());
    return l;
  } else {
    return dbusError(r.error());
  }
}

File: lib/networkmanager.js
/*global exports,  */

/** @module networkmanager */

var dbus = require("dbus");
var Rx = require("rx");

exports.getDevices = function getDevices() {
  return dbus.call(dbus.systemBus,
                   "org.freedesktop.NetworkManager",
                   "/org/freedesktop/NetworkManager",
                   "org.freedesktop.NetworkManager",
                   "GetDevices")
    .flatMap(function(r) {
      return Rx.Observable.fromArray(r[0]);
    });
}

function devInfo(path, iface) {
  return dbus.getProperties(dbus.systemBus, "org.freedesktop.NetworkManager",
                            path, iface)
    .pluck(0);
}

function merge(a, b) {
  var i, out = {};
  for (i in a) {
    if (a.hasOwnProperty(i)) out[i] = a[i];
  }
  for (i in b) {
    if (b.hasOwnProperty(i)) out[i] = b[i];
  }
  return out;
}

function mergeDev(path, iface) {
  return function(dev) {
    return devInfo(path, iface).catch(Rx.Observable.return({})).map(function(newDev) {
      return merge(dev, newDev);
    });
  };
}

exports.getDeviceInfo = function getDeviceInfo(path) {
  return devInfo(path, "org.freedesktop.NetworkManager.Device")
    .map(function(dev) {
      return merge(dev, {objectPath: path});
    })
    .flatMap(mergeDev(path, "org.freedesktop.NetworkManager.Device.Wireless"))
    .flatMap(mergeDev(path, "org.freedesktop.NetworkManager.Device.Wired"))
    .flatMap(mergeDev(path, "org.freedesktop.NetworkManager.Device.Cdma"))
    .flatMap(mergeDev(path, "org.freedesktop.NetworkManager.Device.Gsm"))
    .flatMap(mergeDev(path, "org.freedesktop.NetworkManager.Device.Serial"))
    .flatMap(mergeDev(path, "org.freedesktop.NetworkManager.Device.Bluetooth"))
    .flatMap(mergeDev(path, "org.freedesktop.NetworkManager.Device.OlpcMesh"))
    .flatMap(mergeDev(path, "org.freedesktop.NetworkManager.Device.Modem"));
}
